libs/auth/src/artifact_authenticator.ts

```diff
@@ -0,0 +1,201 @@
+import { Buffer } from 'buffer';
+import fs from 'fs/promises';
+import {
+  assert,
+  err,
+  ok,
+  Result,
+  throwIllegalValue,
+} from '@votingworks/basics';
+
+import { parseCert } from './certs';
+import {
+  constructArtifactAuthenticatorConfig,
+  ArtifactAuthenticatorConfig,
+} from './config';
+import { FileKey, TpmKey } from './keys';
+import {
+  extractPublicKeyFromCert,
+  signMessageUsingPrivateKeyFile,
+  verifyFirstCertWasSignedBySecondCert,
+  verifySignature,
+} from './openssl';
+
+/**
+ * A machine-exported artifact whose authenticity we want to be able to verify
+ */
+export interface Artifact {
+  type: 'cvr_file' | 'election_definition';
+  path: string;
+}
+
+interface SignatureFileContents {
+  signature: Buffer;
+  signingMachineCert: Buffer;
+}
+
+/**
+ * An artifact authenticator that uses digital signatures to verify the authenticity of
+ * machine-exported artifacts, meeting VVSG2 data protection requirements
+ */
+export class ArtifactAuthenticator {
+  private readonly signingMachineCertPath: string;
+  private readonly signingMachinePrivateKey: FileKey | TpmKey;
+  private readonly vxCertAuthorityCertPath: string;
+
+  constructor(
+    // Support specifying a custom config for tests
+    /* istanbul ignore next */
+    input: ArtifactAuthenticatorConfig = constructArtifactAuthenticatorConfig()
+  ) {
+    this.signingMachineCertPath = input.signingMachineCertPath;
+    this.signingMachinePrivateKey = input.signingMachinePrivateKey;
+    this.vxCertAuthorityCertPath = input.vxCertAuthorityCertPath;
+  }
+
+  /**
+   * Writes a signature file for the provided artifact that can later be used to verify the
+   * artifact's authenticity. The signature file is written to the same file path as the artifact,
+   * but with a .vxsig extension, e.g. /path/to/artifact.txt.vxsig.
+   */
+  async writeSignatureFile(artifact: Artifact): Promise<void> {
+    const message = await this.constructMessage(artifact);
+    const messageSignature = await this.signMessage(message);
+    const signingMachineCert = await fs.readFile(this.signingMachineCertPath);
+    const signatureFile = this.serializeSignatureFileContents({
+      signature: messageSignature,
+      signingMachineCert,
+    });
+    await fs.writeFile(
+      this.constructSignatureFilePath(artifact),
+      signatureFile
+    );
+  }
+
+  /**
+   * Verifies the authenticity of the provided artifact using its signature file, which is expected
+   * to be found at the same file path as the artifact, but with a .vxsig extension, e.g.
+   * /path/to/artifact.txt.vxsig.
+   */
+  async authenticateArtifactUsingSignatureFile(
+    artifact: Artifact
+  ): Promise<Result<void, Error>> {
+    try {
+      await this.authenticateArtifactUsingSignatureFileHelper(artifact);
+    } catch {
+      // TODO: Log raw error
```

Comment:
Is this TODO addressed in a later commit?

Reply:
Not yet! I have an issue open for a few logs that I want to add: https://github.com/votingworks/vxsuite/issues/3484